Scene selection for yamaha receivers

stevusprimus opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Yamaha receivers do support the selecting of scenes. Now this can be done via a service call, as the underlying rxv lib already supports this.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests


## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/13673

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

StefanSchornEvorait committed:
Added Service Supprot to select a scene for Yamaha receivers

StefanSchornEvorait committed:
Corrected Service register

homeassistant commented:
Hi @StefanSchornEvorait, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#36488).

Once you do that we will be able to review and accept this pull request.

Thanks!

StefanSchornEvorait committed:
Lint correction

MartinHjelmare commented on homeassistant/components/yamaha/media_player.py:
```diff
@@ -147,7 +151,35 @@ def service_handler(service):
+            device.select_scene(scene)
+            device.schedule_update_ha_state(True)
+
+    hass.services.register(
```
Please refactor and use our entity service helper instead.

https://developers.home-assistant.io/docs/dev_101_services#entity-services

StevusPrimus replied:
Hey, did that. But for that the whole discovery had to be migrated. Learned a lot :)

stevusprimus committed:
Migrated Integration to async handling

stevusprimus committed:
Corrected and added Unit Tests

stevusprimus committed:
Corrected assertion

MartinHjelmare commented on homeassistant/components/yamaha/media_player.py:
```diff
@@ -89,68 +93,62 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+            ctrl_url = f"http://{host}:80/YamahaRemoteControl/ctrl"
+            receivers = rxv.RXV(ctrl_url, name).zone_controllers()
+
+        devices = []
```
Move the code below out of the discovery function.

Please rename `devices` to `entities`.

MartinHjelmare commented on homeassistant/components/yamaha/media_player.py:
```diff
@@ -89,68 +93,62 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+            if receiver.zone in zone_ignore:
+                continue
+
+            device = YamahaDevice(
```
Please rename `device` to `entity`.

MartinHjelmare commented on homeassistant/components/yamaha/media_player.py:
```diff
@@ -89,68 +93,62 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+
+            # Only add device if it's not already added
+            if device.zone_id not in hass.data[DATA_YAMAHA]:
+                hass.data[DATA_YAMAHA][device.zone_id] = device
```
Just store the `zone_id`s in a set. We don't need to store the entities in a dictionary anymore.

MartinHjelmare commented on homeassistant/components/yamaha/media_player.py:
```diff
@@ -89,68 +93,62 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
 
-            device.enable_output(port, enabled)
-            device.schedule_update_ha_state(True)
+    hass.async_add_executor_job(_discovery)
```
`await` the executor job task.
```py
receivers = await hass.async_add_executor_job(_discovery)
```

StevusPrimus replied:
`entities = await hass.async_add_executor_job(_discovery)`

MartinHjelmare commented on homeassistant/components/yamaha/media_player.py:
```diff
@@ -89,68 +93,62 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
-        for device in devices:
-            port = service.data[ATTR_PORT]
-            enabled = service.data[ATTR_ENABLED]
+        hass.add_job(async_add_entities, devices)
```
We'll move this out of discovery function so we don't need to add a job.

MartinHjelmare commented on homeassistant/components/yamaha/media_player.py:
```diff
@@ -368,10 +366,27 @@ def play_media(self, media_type, media_id, **kwargs):
         if media_type == "NET RADIO":
             self.receiver.net_radio(media_id)
 
+    async def async_enable_output(self, enabled, port):
```
We don't need these async wrappers. Our service helper will handle sync methods too.

MartinHjelmare commented on homeassistant/components/yamaha/media_player.py:
```diff
@@ -368,10 +366,27 @@ def play_media(self, media_type, media_id, **kwargs):
+        await self.hass.async_add_job(self.set_scene, scene)
+
+    @property
+    def scene(self):
```
Where is this used?

StevusPrimus replied:
Nowhere, I added it for completeness ...

MartinHjelmare replied:
Please remove it.

StevusPrimus replied:
Removed it as my Receiver does not return the last scene but the source.

MartinHjelmare commented on homeassistant/components/yamaha/media_player.py:
```diff
@@ -89,68 +93,62 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
     if hass.data.get(DATA_YAMAHA) is None:
         hass.data[DATA_YAMAHA] = {}
```
```py
known_zones = hass.data.setdefault(DATA_YAMAHA, set())
```

MartinHjelmare commented on homeassistant/components/yamaha/media_player.py:
```diff
@@ -89,68 +93,62 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+            )
+
+            # Only add device if it's not already added
+            if device.zone_id not in hass.data[DATA_YAMAHA]:
```
```py
if entity.zone_id not in known_zones:
    known_zones.add(entity.zone_id)
```

MartinHjelmare commented on homeassistant/components/yamaha/media_player.py:
```diff
@@ -53,6 +55,8 @@
     {vol.Required(ATTR_ENABLED): cv.boolean, vol.Required(ATTR_PORT): cv.string}
 )
 
+SELECT_SCENE_SCHEMA = MEDIA_PLAYER_SCHEMA.extend({vol.Required(ATTR_SCENE): cv.string})
```
Please inline the schemas when registering the service instead as in the docs example. Leave out the entity_id item.

StevusPrimus replied:
Nowhere, I added it for completeness ...

StevusPrimus commented on homeassistant/components/yamaha/media_player.py:
```diff
@@ -89,68 +93,62 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
-        ]
+            ctrl_url = f"http://{host}:80/YamahaRemoteControl/ctrl"
+            receivers = rxv.RXV(ctrl_url, name).zone_controllers()
+
```
Hey, I am not sure, but I think, it would be more correct to return `devices`?

MartinHjelmare replied:
Do you want to rename `receivers` to `devices`?

We don't want to create the entities inside the discovery function.

StevusPrimus replied:
I return `entitites` now

MartinHjelmare replied:
We should not create the entities inside the discovery function. There should be no side effects when creating the entity so we don't need to do that in the executor.

stevusprimus committed:
Corrections from code review.
https://github.com/home-assistant/core/pull/36488#pullrequestreview-425800517

MartinHjelmare commented on homeassistant/components/yamaha/media_player.py:
```diff
@@ -79,78 +77,74 @@
-        for device in devices:
-            port = service.data[ATTR_PORT]
-            enabled = service.data[ATTR_ENABLED]
+        return entities
```
Please move the entity creation out of the discovery function and return `receivers` here instead.

stevusprimus committed:
Corrections from code review.
https://github.com/home-assistant/core/pull/36488#pullrequestreview-425800517

stevusprimus committed:
Corrections from code review.
https://github.com/home-assistant/core/pull/36488#pullrequestreview-425800517

stevusprimus committed:
Lint corrections

MartinHjelmare commented on homeassistant/components/yamaha/media_player.py:
```diff
@@ -49,10 +51,6 @@
 
 MEDIA_PLAYER_SCHEMA = vol.Schema({ATTR_ENTITY_ID: cv.comp_entity_ids})
```
We can remove this schema now.

MartinHjelmare commented on homeassistant/components/yamaha/media_player.py:
```diff
@@ -79,78 +77,94 @@
+
+    config_info = YamahaConfigInfo(config=config, discovery_info=discovery_info)
+
+    def _discovery(config_info):
```
We can move the discovery function outside `async_setup_platform` now after introducing the config_info parameter. Nice!

stevusprimus committed:
Corrections from code review.
https://github.com/home-assistant/core/pull/36488#pullrequestreview-425800517

Removed not needed constants
Moved the _discovery function out of the platform setup

MartinHjelmare reviewed (approved):
Looks good!

StevusPrimus commented:
I am gonna close this and create a complete new one, as I commited with a wrong account at the beginning which shall not appear later. Thanks for everything

StevusPrimus commented:
https://github.com/home-assistant/core/pull/36564
